frame/support/src/traits/tokens/fungible.rs

```diff
@@ -0,0 +1,218 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2019-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The traits for dealing with a single fungible token class and any associated types.
+
+use super::*;
+use sp_runtime::traits::Saturating;
+use crate::traits::misc::Get;
+use crate::dispatch::{DispatchResult, DispatchError};
+use super::misc::{DepositConsequence, WithdrawConsequence, Balance};
+
+mod balanced;
+mod imbalance;
+pub use balanced::{Balanced, Unbalanced};
+pub use imbalance::{Imbalance, HandleImbalanceDrop, DebtOf, CreditOf};
+
+/// Trait for providing balance-inspection access to a fungible asset.
+pub trait Inspect<AccountId> {
+	/// Scalar type for representing balance of an account.
+	type Balance: Balance;
+	/// The total amount of issuance in the system.
+	fn total_issuance() -> Self::Balance;
+	/// The minimum balance any single account may have.
+	fn minimum_balance() -> Self::Balance;
+	/// Get the balance of `who`.
+	fn balance(who: &AccountId) -> Self::Balance;
+	/// Returns `true` if the balance of `who` may be increased by `amount`.
+	fn can_deposit(who: &AccountId, amount: Self::Balance) -> DepositConsequence;
+	/// Returns `Failed` if the balance of `who` may not be decreased by `amount`, otherwise
+	/// the consequence.
+	fn can_withdraw(who: &AccountId, amount: Self::Balance) -> WithdrawConsequence<Self::Balance>;
+}
+
+/// Trait for providing an ERC-20 style fungible asset.
+pub trait Mutate<AccountId>: Inspect<AccountId> {
+	/// Increase the balance of `who` by `amount`.
+	fn deposit(who: &AccountId, amount: Self::Balance) -> DispatchResult;
+	/// Attempt to reduce the balance of `who` by `amount`.
+	fn withdraw(who: &AccountId, amount: Self::Balance) -> Result<Self::Balance, DispatchError>;
+	/// Transfer funds from one account into another.
+	fn transfer(
+		source: &AccountId,
+		dest: &AccountId,
+		amount: Self::Balance,
+	) -> Result<Self::Balance, DispatchError> {
+		let extra = Self::can_withdraw(&source, amount).into_result()?;
+		Self::can_deposit(&dest, amount + extra).into_result()?;
```

Comment:
Is it okay to use unsafe math in an automatic implementation?

Since we do not implement `can_withdraw`, we cannot guarantee that amount + extra will not overflow.

Of course this would come from a bug or mis-implementation, but still...